mon: prevent pg_temps with size > pool size

liewegas opened the pull request:
http://tracker.ceph.com/issues/26866

Plus some cleanups.

neha-ojha commented on src/osd/OSDMap.cc:
```diff
@@ -1639,18 +1639,17 @@ void OSDMap::dedup(const OSDMap *o, OSDMap *n)
     // it.  we don't care about pg_temps on the pending_inc either; if there
     // are new_pg_temp entries on the pending, clear them out just as well.
-    if (!osdmap.have_pg_pool(pg.first.pool())) {
+    if (!tmpmap.have_pg_pool(pg.first.pool())) {
```
should this be `if (!oldmap.have_pg_pool(pg.first.pool()))`?

liewegas replied:
I think this is an improvement over the previous code because it will remove new pg_temp entries if the pool is simultaneously removed.  The old code would end up with one osdmap epoch with the useless entries and remove them in the next epoch.

xiexingguo commented on src/osd/OSDMap.h:
```diff
@@ -999,15 +999,18 @@ class OSDMap {
   void maybe_remove_pg_upmaps(CephContext *cct,
-                              const OSDMap& osdmap,
+                              const OSDMap& oldmap,
+			      const OSDMap& nextmap,
```
s/nextmap/tmpmap/

xiexingguo commented on src/osd/OSDMap.h:
```diff
@@ -999,15 +999,18 @@ class OSDMap {
-  static void clean_temps(CephContext *cct, const OSDMap& osdmap,
+  static void clean_temps(CephContext *cct,
+			  const OSDMap& oldmap,
+			  const OSDMap& nextmap,
```
s/nextmap/tmpmap/

xiexingguo commented on src/osd/OSDMap.cc:
```diff
@@ -1681,6 +1681,15 @@ void OSDMap::clean_temps(CephContext *cct,
+      ldout(cct, 10) << __func__ << "  removing pg_temp " << pg.first << " "
+		     << pg.second << " exceeds pool size" << dendl;
+      if (oldmap.pg_temp->count(pg.first))
+	pending_inc->new_pg_temp[pg.first].clear();
```
too much dup code...

liewegas commented:
better?

gregsfortytwo commented on src/osd/OSDMap.cc:
```diff
@@ -1673,9 +1673,19 @@ void OSDMap::clean_temps(CephContext *cct,
+    if (pg.second.size() > tmpmap.get_pg_pool(pg.first.pool())->get_size()) {
+      ldout(cct, 10) << __func__ << "  removing pg_temp " << pg.first << " "
+		     << pg.second << " exceeds pool size" << dendl;
+      remove = true;
```
I’m a bit worried that this invites pg_temp create/destroy loops as this code is divorced from that which sets them in the OSD.

liewegas commented:
This is responsible for http://pulpito.ceph.com/sage-2018-08-09_03:13:30-rados:standalone-wip-sage-testing-2018-08-08-1056-distro-basic-smithi/

dzafman commented:
The test code is getting confused because it assumes that only a single OSD performs backfill.  For some reason backfill started on osd.1 "[1,2,3]/[1,2,0] backfill=[3]" then happens on osd.0 " [3]/[0,1] backfill=[3]".  The test sets "nobackfill" before making a bunch of changes to cause a single backfill to occur.

```
2018-08-14` 15:02:46.085 7fa3e4db5700 20 osd.1 pg_epoch: 25 pg[1.0( v 21'200 (21'100,21'200] local-lis/les=24/25 n=200 ec=18/18 lis/c 24/18 les/c/f 25/19/0 23/24/18) [1,2,3]/[1,2,0] backfill=[3] r=0 lpr=24 pi=[18,24)/1 bft=3 crt=21'200 lcod 21'199 mlcod 0'0 active+remapped+backfilling mbc={}] _update_calc_stats actingset 0,1,2 upset 1,2,3 acting_recovery_backfill 0,1,2,3
```

Has nobackfill handling changes, so that the pg state can go into backfilling even though no progress is actually made?

This fixes the test to detect this error:

```$ git diff
diff --git a/qa/standalone/osd/osd-backfill-stats.sh b/qa/standalone/osd/osd-backfill-stats.sh
index d57305f..7999028 100755
--- a/qa/standalone/osd/osd-backfill-stats.sh
+++ b/qa/standalone/osd/osd-backfill-stats.sh
@@ -73,6 +73,11 @@ function check() {

     local log=$(grep -l +backfilling $dir/osd.*.log)
     test -n "$log" || return 1
+    if [ "$(echo "$log" | wc -w)" != "1" ];
+    then
+      echo "Test setup failure, a single OSD should have performed backfill"
+      return 1
+    fi

     local addp=" "
     if [ "$type" = "erasure" ];
```

liewegas commented:
THere's no change except for the pg_temp behavior.  I instrumented the test to dump every osdmap, and the only change is that in one epoch the pg_temp is removed at the same time pool size is set to 2.  And thereafter the new pg_temp has a different set of osds (1 is swapped w/ 0 iirc).

Try reverting that one commit and everything will work again.

So i'm not quite sure what is going on...

dzafman commented:
@liewegas Excluding 90764e94fd327b273b1192e6b775faab0dd5febb causes osd-backfill-status.sh to pass.

Cherry-pick https://github.com/dzafman/ceph/commit/1ca2c048c6ea06062d1b3a12d2ed5fb353d7bd27 to get the test improvement.

dzafman commented:
@liewegas The test change issues a message about the reason for a test failure, but the test still fails.  Without 
https://github.com/dzafman/ceph/commit/1ca2c048c6ea06062d1b3a12d2ed5fb353d7bd27 all sub-tests pass in particular TEST_backfill_remapped which was broken.

The tests require that changes inside set nobackfill/unset nobackfill will cause cause a single backfill event on the new primary.

dzafman commented:
jenkins test docs please

liewegas commented:
@dzafman Ok, I've fixed the test.  Two things happened:

- a side-effect of the change is that the primary changes when the pool size is changed.  the test now recalculates the primary at the appropriate time.
- the check() function was ignoring the primary and checking all OSDs' logs, under the assumption that the primary didn't change.  Fixed.

It passes now!
